gordon/metrics/log.py

```diff
@@ -0,0 +1,123 @@
+# limitations under the License.
+
+import logging
+import time
+
+import zope.interface
+
+from gordon import interfaces
+
+# TODO: update plugin loader to load metrics logger as default
+
+
+@zope.interface.implementer(interfaces.ITimer)
+class LogTimer:
+    """Timer that outputs metrics to the logger.
+
+    Args:
+        metric (dict): Metric to output.
+        logger (.MetricLogger): Object used for outputting metrics.
+        time_unit (number): (optional) Multiplier used to scale time.time()
+            output.
+    """
+    def __init__(self, metric, logger, time_unit=1):
+        self.metric = metric
+        self.logger = logger
+        self.time_unit = time_unit
+
+    async def __aenter__(self):
+        await self.start()
+        return self
+
+    async def __aexit__(self, *args):
+        await self.stop()
+
+    async def start(self):
+        """Start the timer.
+        """
+        self._start_time = time.time()
+
+    async def stop(self):
+        """Stop the timer and output the time delta to the logger.
+        """
+        time_elapsed = (time.time() - self._start_time) * self.time_unit
+        self.metric['value'] = time_elapsed
+        self.logger.log(self.metric)
+
+
+class MetricLogger:
+    """Configurable wrapper that formats and outputs metrics.
+
+    Args:
+        level (str): Log level at which to output metric.
+        precision (str): Used to round metric to desired number of
+            places.
+    """
+    LOGFMT = '[{metric_name}]\tvalue: {value:.<prec>f}\tattrs: {attrs}'
+
+    def __init__(self, level, precision):
+        self._log_fmt = self.LOGFMT.replace('<prec>', str(precision))
+        self._log_fn = getattr(logging, level.lower())
+
+    def log(self, metric):
+        """Format and output metric.
+
+        Args:
+            metric (dict): Complete metric.
+        """
+        message = self._log_fmt.format(**metric)
+        self._log_fn(message)
+
+
+@zope.interface.implementer(interfaces.IMetricRelay)
+class LogRelay:
+    """Manager metrics that get output to the logger.
+
+    Args:
+        config (dict): Required keys:
+            level (str): Log level of metrics log messages.
+            precision (int): Number of decimal places to round metric
+                value to for output.
+            time_unit (number): Scale unit to use with timers.
+    """
+    def __init__(self, config):
+        level = config.get('log_level', 'info')
+        precision = config.get('precision', 6)
+        self.logger = MetricLogger(level, precision)
+        self.time_unit = config['time_unit']
+
+    def _create_metric(self, metric_name, value, **kwargs):
+        attrs = kwargs.get('attrs') or {}
+
+        return {
+            'metric_name': metric_name,
+            'value': value,
+            'attrs': attrs
+        }
+
+    async def incr(self, metric_name, value=1, **kwargs):
+        await self.set(metric_name, value, **kwargs)
```

Comment:
Hmm, this wouldn't be terribly useful. Do you think there should be a stored value somewhere? i.e.

```python
async def incr(self, metric_name, value=1, **kwargs):
    self.counters['metric_name'] += value
    await self.set(metric_name, self.counters['metric_name'], **kwargs)
```


Output something like:

```
[router-msg-consumed] value: 10 attrs: ...
```

Rather than:
```
[router-msg-consumed] value: 1 attrs: ...
```

Reply:
:fox_face: Excellent point. I'll rework this bit to accumulate values.